### DISCUSSION TITLE
Wish List: Build numbers

### DISCUSSION THREAD

**Discussion by runette** [openupm/openupm#1906] [Ideas]
Sometimes:

- You are integrating source into UPM that also have an existence in other domains (i.e. they have a version number that is not fully controlled by the people creating the UPM package).

- You want, in this case, for the UPM package version to reflect the wider version number for the source to make it clear.

However, if you make a mistake in the package metadata you cannot replace the package with a new version with the same version number.

I WISH I HAD: A "build number" as well as a version in `package.json` that could be used to signal to OpenUPM that this is an update to the package metadata only and does not change the ABI for the package - so that e.g. build 5 of version 1.0.3 is source/binary compatible with build 4 of version 1.0.3 and can directly replace it safely and thus errors in metadata can be corrected.

Version bumps should be used for changes that change the ABI and not for changes to the metadata.

This is the solution that the Conda repository uses for this problem and it works very well.

**Comment by favoyang** [comment]
Semver 2.0.0 supports build number using the "+" operator. e.g. `1.0.3+004`. But the build number is ignored when publishing to a registry. https://github.com/npm/cli/issues/1479

So submitting a new package `1.0.3+005` to a registry with version `1.0.3` (actually `1.0.3-004`) will get version conflicts.

I understand in your case perhaps a mistake of metadata or the build receipt changed that you want to publish a new build. But it's not possible to be implemented in an NPM registry unless we decide to replace the previous build with the same version - that's something OpenUPM is trying to avoid - to unpublish published content.

Perhaps a workaround is to use a version pattern documented in https://medium.com/openupm/how-to-maintain-upm-package-part-3-2d08294269ad#88d8

**Comment by runette** [comment]
>
> unless we decide to replace the previous build with the same version -
> that's something OpenUPM is trying to avoid - to unpublish published
> content.


That is precisely what Conda allows - trusting the package publisher to do
the right thing. If the ABI is unchanged (i.e. the update has no effect at
all on the consuming code) then there is no problem with replacing the
package. The problems come from removing a package or replacing a package
with one with a different ABI.

However, the version pattern you mentioned looks like a possible workaround
while at the same time my "wish list" remains a wish.

**Reply by favoyang** [reply]
> That is precisely what Conda allows - trusting the package publisher to do the right thing.

I think you're right that the Conda server *replaced* the old build with the newer one.

```
# Find a version with more than one build
$ conda search pillow | grep -v _0
pillow                         4.3.0  py27h353bd0c_1  pkgs/main
pillow                         4.3.0  py35h550890c_1  pkgs/main
pillow                         4.3.0  py36h6f462bf_1  pkgs/main

# Install the previous build, expected to fail
$ conda install pillow=4.3.0=py36h6f462bf_0
...
PackagesNotFoundError: The following packages are not available from current channels:
  - pillow==4.3.0=py36h6f462bf_0
```

That's an interesting design decision. Because Conda is basically a re-packing system it facing the exact issue you mentioned, and conda also documented in https://conda.io/projects/conda-build/en/latest/resources/package-spec.html, the field is useful to deal with bugs introduced by build receipt. 

```
build_number

A non-negative integer representing the build number of the package.

...This is important because new builds that contain bug fixes for the way a package is built may be added to a repository.
```

It can be abused to create another layer of version control that can not notify end-users who installed an old build. Based on my experience with this project, package owners are not all version control experts. More than one time I'm been asked to rebuild a specific version because they messed up something and want to keep the version sequence "clean" with a re-build.

I didn't completely reject this request. I can see its value if it's been carefully implemented. Even though I need to research more on the registry implementation to verify how it deals with the build number. The filename is one example, Conda has the build number as part of the tarball filename (while NPM didn't have it), so it won't mess up the CDN/Edge server, or have to invalidate it.

For now, I suggest you evaluate the workarounds, either
- stay with the pre-release stage forever: `1.0.1` => `1.0.1-b0` (build 0), `1.0.1-b1` (build 1)
- or, use the scaled version solution: `1.0.1` => `1.0.100` (build 0), `1.0.101` (build 1)

**Comment by runette** [comment]
No. Won't work. Cannot cope with x.x.0

**Reply by favoyang** [reply]
As mentioned in the article,

> It works with one “glitch” when the upstream patch number is zero. If the original version is 3.0.0 your version will be 3.0.1, 3.0.2... Because 3.0.001 transforms into 3.0.1 during the npm publishing. It can be slightly confusing but won’t be conflicted.

| Original version | Scaled original version | NPM actual Version | Expected patch version | NPM actual patch version |
|-|-|-|-|-|
| 3.0.0 | 3.0.000 | 3.0.0 | 3.0.001 | 3.0.1 |
| 3.0.1 | 3.0.100 | 3.0.100 | 3.0.101 | 3.0.101 |

**Comment by runette** [comment]
I find a system that turns source version "3.0.0"  into package version
"3.0.1" when the next source version would be "3.0.1" counter-intuitive -
especially when in an effect analogous to Benford's Law the semver pattern
x.x.0 is actually the most common pattern.

**Reply by favoyang** [reply]
That's true. It's working but far from perfect.

I want to point out a common case that the migration author needs to write an adopter layer to migrate a library from other domains into a upm format. The adopter layer is also part of the "source". For this case, the build number is not useful, you have to figure out a meaningful way to mix your version and the upstream version. That's the background of inventing the scaled version solution.

**Comment by runette** [comment]
You are right that such a system does put more trust in package owners. But
then - package owners CAN break a package now if they want to - just not in
this way.

And I understand that it is a complicated matter. If the problem is in npm
then it may not possible.

Both of those are why I raised it as a "wish item" and nothing stronger. I
do believe that such a mechanism would make the repository stronger and
also reduce the load on you with such requests - as well as increasing the
projects "bus factor".

**Comment by runette** [comment]
Understood - I still think a system that converts 3.0.0 to 3.0.1 and 3.0.1
to 3.0.101 has a critical flaw. That said I have nothing better at the
moment.

Interestingly - a quick Google search shows that this issue has been a
background complaint about NPM for many years (e.g.
https://gist.github.com/JamesMGreene/9916485). If only they had just
provided a solution rather than trying to be "morally pure" about it!

It brings to mind the old English saying about the shopkeeper who says to a
customer who asked for a product "How many times do I have to tell
customers that there is no demand for that product!".
